tests/integration/plusminus/plusminus/BaseCalculator.py

```diff
@@ -0,0 +1,243 @@
+            )
+
+    @property
+    def output_filenames(self):
+        """Native calculator file names"""
+        return self.__output_filenames
+
+    @output_filenames.setter
+    def output_filenames(self, value):
+        self.set_output_filenames(value)
+
+    def set_output_filenames(self, value: Union[list, str]):
+        if isinstance(value, str):
+            self.__output_filenames = [value]
+        elif isinstance(value, list):
+            self.__output_filenames = value
+        else:
+            raise TypeError(
+                f"Calculator: `output_filenames` can to be a str or a list, and will be treated as a list, but not {type(value)}"
+            )
+
+    @property
+    def instrument_base_dir(self):
+        return self.__instrument_base_dir
+
+    @instrument_base_dir.setter
+    def instrument_base_dir(self, value):
+        self.set_instrument_base_dir(value)
+
+    def set_instrument_base_dir(self, value: str):
+        if isinstance(value, str):
+            self.__instrument_base_dir = value
+        else:
+            raise TypeError(
+                f"Calculator: `instrument_base_dir` is expected to be a str, not {type(value)}"
+            )
+
+    @property
+    def calculator_base_dir(self):
+        return self.__calculator_base_dir
+
+    @calculator_base_dir.setter
+    def calculator_base_dir(self, value):
+        self.set_calculator_base_dir(value)
+
+    def set_calculator_base_dir(self, value: str):
+        if isinstance(value, str):
+            self.__calculator_base_dir = value
+        else:
+            raise TypeError(
+                f"Calculator: `calculator_base_dir` is expected to be a str, not {type(value)}"
+            )
+
+    @property
+    def base_dir(self):
+        base_dir = Path(self.instrument_base_dir) / self.calculator_base_dir
+        return str(base_dir)
+
+    @property
+    def output_file_paths(self):
+        paths = []
+        for filename in self.output_filenames:
+            path = Path(self.base_dir) / filename
+            # Make sure the file directory exists
+            path.parent.mkdir(parents=True, exist_ok=True)
+            paths.append(str(path))
+        return paths
+
+    @abstractmethod
+    def backengine(self):
+        Path(self.base_dir).mkdir(parents=True, exist_ok=True)
+        input_num0 = self.input[self.input_keys[0]].get_data()["number"]
+        input_num1 = self.input[self.input_keys[1]].get_data()["number"]
+        output_num = float(input_num0) + float(input_num1)
+        if self.parameters["plus_times"].value > 1:
+            for i in range(self.parameters["plus_times"].value - 1):
+                output_num += input_num1
+        data_dict = {"number": output_num}
+        key = self.output_keys[0]
+        output_data = NumberData.from_dict(data_dict, key)
+        self.output = DataCollection(output_data)
+        return self.output
```

Comment:
I was under the impression that backengine calls should not return anything, but instead place the data in calculator.data? I like backengine returning results, but was okay with following libpyvinyl convention.

Reply:
Yes, there was back-and-forth in that topic. In the end, I think we decided to return the results in the `backengine()`. Nevertheless, we didn't see any problem in the tests.